### PR TITLE
Improve organization listing service

### PR DESCRIPTION
**Pull request by inthirakumaaran**
## Purpose
- Remove the permission restriction in the default listing service
- Introduce a new service with user permission restriction
- Introduce a new method to check for the uniqueness of the organization name 

Ref: https://github.com/wso2-enterprise/asgardeo-product/issues/17262

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/dao/impl/OrganizationManagementDAOImpl.java [hunk 1]
```diff
@@ -699,6 +636,95 @@ public String resolveTenantDomain(String organizationId) throws OrganizationMana
...
+        Map<String, String> filterAttributeValue = filterQueryBuilder.getFilterAttributeValue();
+        List<String> timestampTypeAttributes = filterQueryBuilder.getTimestampFilterAttributes();
+
+        FilterQueryBuilder parentIdFilterQueryBuilder = new FilterQueryBuilder();
+        appendFilterQueryForParentId(parentIdFilterQueryBuilder, parentIdExpressionNodes);
+        Map<String, String> parentIdFilterAttributeValueMap = parentIdFilterQueryBuilder.getFilterAttributeValue();
+        String parentIdFilterQuery = parentIdFilterQueryBuilder.getFilterQuery();
+
+        String sqlStmt;
+        String getOrgSqlStmtTail = GET_ORGANIZATIONS_TAIL;
```
Comment (ravinduramesh) [inline components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/dao/impl/OrganizationManagementDAOImpl.java:656]: Can we use this query itself if `allSubOrgs = true` and the DB is MySQL?

File: components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/dao/OrganizationManagementDAO.java [hunk 2]
```diff
@@ -105,6 +105,24 @@ List<BasicOrganization> getOrganizations(boolean recursive, Integer limit, Strin
...
+     * @param recursive               Determines whether records should be retrieved in a recursive manner.
+     * @param limit                   The maximum number of records to be returned.
+     * @param organizationId          The organization ID.
+     * @param sortOrder               The sort order, ascending or descending.
+     * @param expressionNodes         The list of filters excluding filtering by parentId.
+     * @param parentIdExpressionNodes The list of filters related to parentId.
+     * @return the list of organization IDs.
+     * @throws OrganizationManagementServerException The server exception thrown when retrieving the organizations.
+     */
+    List<BasicOrganization> getOrganizationsByUser(boolean recursive, Integer limit, String organizationId,
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/dao/OrganizationManagementDAO.java:121]: suggestion : `getUserAuthorizedOrganizations`

File: components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/dao/impl/OrganizationManagementDAOImpl.java [hunk 3]
```diff
@@ -699,6 +636,95 @@ public String resolveTenantDomain(String organizationId) throws OrganizationMana
         }
     }
 
+    private List<BasicOrganization> getOrganizationsList(boolean allSubOrgs, boolean recursive, Integer limit,
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/dao/impl/OrganizationManagementDAOImpl.java:639]: suggestion to `allSubOrgs`
AFAIU, that variable is to indicate whether org list should be returned based on user permissions or not
suggestion: `authorizedSubOrgsOnly`/ `includeSubOrgsByPermission` / `allowedSubOrgsOnly`

File: components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/dao/impl/OrganizationManagementDAOImpl.java [hunk 4]
```diff
@@ -699,6 +636,95 @@ public String resolveTenantDomain(String organizationId) throws OrganizationMana
...
+
+        String permissionPlaceholder = "PERMISSION_";
+        List<String> permissions = getAllowedPermissions(VIEW_ORGANIZATION_PERMISSION);
+        List<String> permissionPlaceholders = new ArrayList<>();
+        // Constructing the placeholders required to hold the permission strings in the named prepared statement.
+        for (int i = 1; i <= permissions.size(); i++) {
+            permissionPlaceholders.add(":" + permissionPlaceholder + i + ";");
+        }
+        String placeholder = String.join(", ", permissionPlaceholders);
+        sqlStmt = sqlStmt.replace(PERMISSION_LIST_PLACEHOLDER, placeholder);
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/dao/impl/OrganizationManagementDAOImpl.java:683]: is this required if allSubOrgs == true ?

File: components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/dao/impl/OrganizationManagementDAOImpl.java (hunk 1 above)
Comment (inthirakumaaran) [reply components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/dao/impl/OrganizationManagementDAOImpl.java:656]: I missed this DB, will check on this

File: components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/dao/impl/OrganizationManagementDAOImpl.java (hunk 4 above)
Comment (inthirakumaaran) [reply components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/dao/impl/OrganizationManagementDAOImpl.java:683]: we can stop them as well

**Review by thanujalk — APPROVED** [review]

**Pull request merged**